app/main.py

```diff
@@ -207,12 +207,30 @@ def get_repo_options(request, lang):
                                     test_mode=test_mode))
     return options
 
-def get_language_options(request, config=None):
+def get_language_options(request, config, current_lang):
     """Returns a list of information needed to generate the language menu."""
-    return [{'lang': lang,
-             'endonym': const.LANGUAGE_ENDONYMS.get(lang, '?'),
-             'url': utils.set_url_param(request.url, 'lang', lang)}
-            for lang in (config and config.language_menu_options or ['en'])]
+    primary_langs = (config and config.language_menu_options) or ['en']
+    all_langs = sorted(
+        const.LANGUAGE_ENDONYMS.keys(),
+        key=lambda s: const.LANGUAGE_ENDONYMS[s])
+    return {
+        'primary':
+            [get_language_option(request, lang, lang == current_lang)
+             for lang in primary_langs],
+        'all':
+            [get_language_option(
+                request, lang,
+                lang == current_lang and lang not in primary_langs)
```

Comment:
Added a comment:
+            # We put both 'primary' and 'all' languages into a single <select>
+            # box (See app/resources/language-menu.html.template).
+            # If current_lang is in the primary languages, we mark the
+            # language as is_selected in 'primary', not in 'all', to make sure
+            # a single option is selected in the <select> box.

Also renamed is_current to is_selected.